codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -307,38 +338,87 @@ private Http2Stream requireStream(int streamId) {
         return stream;
     }
 
-    /**
-     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
-     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
-     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
-     * <p>
-     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
-     * the passed promise is not completed until last frame write.
-     * </p>
-     */
-    private final class FlowControlledData extends FlowControlledBase {
-        private ByteBuf data;
-        private int size;
-
-        private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf buf, int padding,
-                                    boolean endOfStream, ChannelPromise promise) {
-            super(ctx, stream, padding, endOfStream, promise);
-            this.data = buf;
-            size = data.readableBytes() + padding;
+    private final class FlowControlledFileRegion extends FlowControlledData<FileRegion> {
+        
+        private FlowControlledFileRegion(ChannelHandlerContext ctx, Http2Stream stream,
+                FileRegion data, int padding, boolean endOfStream, ChannelPromise promise) {
+            super(ctx, stream, data, (int) data.count(), padding, endOfStream, promise);
         }
 
         @Override
-        public int size() {
-            return size;
+        public void write(int allowedBytes) {
+            int bytesWritten = 0;
+            if (data == null || (allowedBytes == 0 && size != 0)) {
+                // No point writing an empty DATA frame, wait for a bigger allowance.
+                return;
+            }
+            try {
+                int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+                do {
+                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                    if (data == null) {
```

Comment:
padding is also considered by flow controller, so it is possible that we finish writing actual data out(then I set data to null in the code), but left some padding bytes behind. So if data == null then we know that we only need to write padding out(thus an empty buffer).